### PR TITLE
NIMBUS-23:: Support for charts + NIM-21748::Project sub document attributes in lookUp queries 

### PR DESCRIPTION
**Pull request by sandeepmantha**
# Description
Project sub document attributes in look up queries
Support for charts

# Overview of Changes
Changed getDynamicParamValues method to use Spel expression 
Added a new component that can render Bar, Line, Pie and Doughnut Charts

# Type of Change
- [ ] New feature

# Test Details
test_dynamicLookUpWithProjection
chart.component.spec.ts

### CODE REVIEW

**Comment by openanthem-admin** [comment]
You did it @sandeepmantha!

Thank you for signing the Anthem Opensource Software CLA.

File: nimbus-core/src/main/java/com/antheminc/oss/nimbus/domain/cmd/exec/internal/search/DefaultSearchFunctionHandlerLookup.java [hunk 1]
```diff
@@ -93,24 +100,57 @@ private R getStaticParamValues(List<StaticCodeValue> searchResult, Command cmd)
...
+//				}
+//				paramValues.add(new ParamValue(codePd.getReadMethod().invoke(model).toString(),labelPd.getReadMethod().invoke(mod).toString()));
+//			}
+//			return (R)paramValues;
+//		}
+//		catch(Exception ex) {
+//			throw new FrameworkRuntimeException("Failed to execute read on property: "+codePd+" and "+labelPd, ex);
+//		}
+//	}
+
```
Comment (lopezton-anthem) [inline nimbus-core/src/main/java/com/antheminc/oss/nimbus/domain/cmd/exec/internal/search/DefaultSearchFunctionHandlerLookup.java]: Per sonar rules, we should remove commented code

File: nimbus-core/src/main/java/com/antheminc/oss/nimbus/domain/cmd/exec/internal/search/DefaultSearchFunctionHandlerLookup.java [hunk 2]
```diff
@@ -93,24 +100,57 @@ private R getStaticParamValues(List<StaticCodeValue> searchResult, Command cmd)
...
 		if(list.size() > 2)
-			throw new IllegalStateException("ParamValues lookup failed due to more than 2 fields provided to create the param values. the criteria class is "+criteriaClass);
-		
-		PropertyDescriptor codePd = BeanUtils.getPropertyDescriptor(criteriaClass, list.get(0));
-		PropertyDescriptor labelPd = BeanUtils.getPropertyDescriptor(criteriaClass, list.get(1));
+		throw new IllegalStateException("ParamValues lookup failed due to more than 2 fields provided to create the param values. the criteria class is "+criteriaClass);
 		
+		String cd = list.get(0).replaceAll("\\.", "\\?.");
+		String lb = list.get(1).replaceAll("\\.", "\\?.");
```
Comment (lopezton-anthem) [inline nimbus-core/src/main/java/com/antheminc/oss/nimbus/domain/cmd/exec/internal/search/DefaultSearchFunctionHandlerLookup.java]: Duplicated constants `"\\."` and `"\\?."`. Consider externalizing or adding these as a static constants.

File: nimbus-test/src/test/java/com/antheminc/oss/nimbus/domain/model/state/builder/DefaultActionExecutorSearchTest.java [hunk 3]
```diff
@@ -202,13 +204,61 @@ public void t5_testSearchByQueryWithProjection() {
...
+		
+		MultiOutput multiOp1 = this.commandGateway.execute(cmdMsg1);
+		List<Output<?>> ops1  = multiOp1.getOutputs();
+		
+		assertNotNull(ops1);
+		
+		List<ParamValue> values1 = (List<ParamValue>)ops1.get(0).getValue();
+		
+		assertNotNull(values1);
+		assertEquals(2, values1.size());
```
Comment (lopezton-anthem) [inline nimbus-test/src/test/java/com/antheminc/oss/nimbus/domain/model/state/builder/DefaultActionExecutorSearchTest.java]: Validate code and label are as expected

File: nimbus-core/src/main/java/com/antheminc/oss/nimbus/domain/cmd/exec/internal/search/DefaultSearchFunctionHandlerLookup.java [hunk 4]
```diff
@@ -93,24 +100,57 @@ private R getStaticParamValues(List<StaticCodeValue> searchResult, Command cmd)
...
+				Object label = expressionParser.parseExpression(lb).getValue(context);
+				if(code!= null && label !=null) {
+					paramValues.add(new ParamValue(code.toString(), label.toString()));
+				}
 			}
 			return (R)paramValues;
 		}
 		catch(Exception ex) {
-			throw new FrameworkRuntimeException("Failed to execute read on property: "+codePd+" and "+labelPd, ex);
+			throw new FrameworkRuntimeException("Failed to parse property - code: "+list.get(0)+" and label: "+list.get(1), ex);
```
Comment (lopezton-anthem) [inline nimbus-core/src/main/java/com/antheminc/oss/nimbus/domain/cmd/exec/internal/search/DefaultSearchFunctionHandlerLookup.java]: Please add a test case to capture the negative scenario here.

File: nimbus-core/src/main/java/com/antheminc/oss/nimbus/domain/cmd/exec/internal/search/DefaultSearchFunctionHandlerLookup.java [hunk 5]
```diff
@@ -95,22 +99,28 @@ private R getStaticParamValues(List<StaticCodeValue> searchResult, Command cmd)
...
-		PropertyDescriptor labelPd = BeanUtils.getPropertyDescriptor(criteriaClass, list.get(1));
+		throw new IllegalStateException("ParamValues lookup failed due to more than 2 fields provided to create the param values. the criteria class is "+criteriaClass);
 		
+		String cd = list.get(0).replaceAll(toReplace, replaceWith);
+		String lb = list.get(1).replaceAll(toReplace, replaceWith);
 		try {
 			List<ParamValue> paramValues = new ArrayList<>();
 			for(Object model: searchResult) {
-				paramValues.add(new ParamValue(codePd.getReadMethod().invoke(model).toString(), (String)labelPd.getReadMethod().invoke(model)));
+				StandardEvaluationContext context = new StandardEvaluationContext(model);
```
Comment (rakeshnpatel) [inline nimbus-core/src/main/java/com/antheminc/oss/nimbus/domain/cmd/exec/internal/search/DefaultSearchFunctionHandlerLookup.java]: instead of instantiating the SpEL context here, we could use the **ExpressionEvaluator** interface which has the implementation for SpEL (**SpelExpressionEvaluator**). Inject this dependency in **DefaultSearchFunctionHandlerLookup** and invoke the getValue methods passing required parameters.

File: nimbus-core/src/main/java/com/antheminc/oss/nimbus/domain/defn/ViewConfig.java [hunk 6]
```diff
@@ -868,7 +868,7 @@
 	@ViewStyle
 	public @interface FilterButton {
 		String alias() default "FilterButton";
-
+Chart
```
Comment (lopezton-anthem) [inline nimbus-core/src/main/java/com/antheminc/oss/nimbus/domain/defn/ViewConfig.java]: Is this a typo?

Comment (sandeepmantha) [reply nimbus-core/src/main/java/com/antheminc/oss/nimbus/domain/defn/ViewConfig.java]: yes, it was.

**Comment by soham-anthem** [comment]
@sandeepmantha In JSON Serializer, the condition to detect if its Grid/Chart would inhibit extension of the framework - just like how you had to make a change to the Serializer to add a new UI behavior. I would suggest to look to an alternate approach that doesnt cause making changes to Serializer for such UI needs. Not stopping the PR, but would request to create a low effort story (1 point) to make the change.
